File: backend/app/models/body_composition.py
"""Body composition data models."""

from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class BodyCompositionBase(BaseModel):
    """Base body composition model."""
    timestamp: datetime
    date: str = Field(..., description="Date in YYYY-MM-DD format")
    weight: float
    weight_unit: str = "lbs"
    body_fat_pct: float | None = Field(None, ge=0, le=100)
    muscle_mass: float | None = None
    bmi: float | None = None
    water_pct: float | None = Field(None, ge=0, le=100)
    # kg, unlike its neighbours, and the name says so. `metric_def` defines
    # bone as kg for DEXA (Plan 0008) and openScale reports kg; a second name
    # in pounds would put one quantity under two names, which is what
    # ADR-0003's naming rule exists to prevent.
    bone_mass_kg: float | None = None
    visceral_fat: float | None = None
    metabolic_age: int | None = None
    protein_pct: float | None = Field(None, ge=0, le=100)


class BodyCompositionCreate(BodyCompositionBase):
    """Model for creating a body composition measurement."""
    pass


class BodyComposition(BodyCompositionBase):
    """Full body composition model with metadata."""
    id: int = Field(..., alias="doc_id")
    created_at: datetime
    # Which instrument produced this. A bioimpedance estimate and a DEXA
    # measurement of the same quantity are not interchangeable, and a consumer
    # that cannot tell them apart will eventually average them.
    source: str = "openscale"

    model_config = ConfigDict(populate_by_name=True, from_attributes=True)


class BodyCompositionStats(BaseModel):
    """Body composition summary statistics.

    The `latest_*` figures are the most recent values recorded, whatever
    produced them. The `*_change` figures are deltas *within a single source* -
    subtracting a DEXA body fat from a bioimpedance one measures the gap
    between two instruments, not a change in the body. `primary_source` names
    the series the deltas describe.
    """
    total_measurements: int
    latest_weight: float | None = None
    latest_body_fat: float | None = None
    latest_muscle_mass: float | None = None
    latest_source: str | None = None
    weight_change: float | None = None
    body_fat_change: float | None = None
    muscle_mass_change: float | None = None
    primary_source: str | None = None
    first_date: str | None = None
    latest_date: str | None = None


class BodyCompositionSyncResult(BaseModel):
    """What one BodySpec sync did.

    `skipped` is the number of scans already held. On a second run it should
    equal `scans_found` with `imported` at zero - idempotency is the property
    most likely to break here, and the failure mode is silent duplicate
    history.
    """
    scans_found: int
    imported: int
    skipped: int
    metrics_written: int


class BodyCompositionTrend(BaseModel):
    """Trend data for charts.

    `sources` runs parallel to `dates`, so a chart can split the points by
    instrument. Without it a client joins a quarterly DEXA point to a daily
    scale reading with a straight line, asserting a trajectory across three
    months that nothing measured.
    """
    dates: list[str]
    weights: list[float | None]
    body_fat_pcts: list[float | None]
    muscle_masses: list[float | None]
    water_pcts: list[float | None]
    sources: list[str] = []

File: frontend/src/types/bodyComposition.ts
export interface BodyComposition {
  doc_id: number;
  timestamp: string;
  date: string;
  weight: number;
  weight_unit: string;
  body_fat_pct: number | null;
  muscle_mass: number | null;
  bmi: number | null;
  water_pct: number | null;
  /**
   * Kilograms, unlike its neighbours — the unit is in the name because
   * `metric_def` defines bone as kg for DEXA and openScale reports kg. See
   * docs/plans/0010-retire-body-composition.md §2.
   */
  bone_mass_kg: number | null;
  visceral_fat: number | null;
  metabolic_age: number | null;
  protein_pct: number | null;
  created_at: string;
  /** Which instrument produced this - 'openscale', 'bodyspec' or 'manual'.
   *  A bioimpedance estimate and a DEXA measurement are not interchangeable. */
  source: string;
}

export interface BodyCompositionStats {
  total_measurements: number;
  latest_weight: number | null;
  latest_body_fat: number | null;
  latest_muscle_mass: number | null;
  /** Source of the most recent measurement the latest_* figures come from. */
  latest_source: string | null;
  weight_change: number | null;
  body_fat_change: number | null;
  muscle_mass_change: number | null;
  /** The single series the *_change figures describe. Deltas are never
   *  computed across instruments - that would report the gap between them
   *  as a change in the body. */
  primary_source: string | null;
  first_date: string | null;
  latest_date: string | null;
}

export interface BodyCompositionTrend {
  dates: string[];
  weights: (number | null)[];
  body_fat_pcts: (number | null)[];
  muscle_masses: (number | null)[];
  water_pcts: (number | null)[];
  /** Runs parallel to `dates`. Without it a chart joins a quarterly DEXA
   *  point to a daily scale reading with a line, asserting a trajectory
   *  across three months that nothing measured. */
  sources: string[];
}

export interface BodySpecSyncResult {
  scans_found: number;
  imported: number;
  skipped: number;
  metrics_written: number;
}

/**
 * What one scale drain did — plan 0015 §4.
 *
 * `skipped` is the number already held. Unlike the BodySpec sync, a high
 * `skipped` is the *normal* result here: the scale replays all thirty stored
 * weighings on every connect, so most of any drain is already in the database.
 */
export interface ScaleSyncResult {
    readings_received: number;
    imported: number;
    skipped: number;
}
